// @arwes/react-frames/build/cjs/FrameSVGNefrex/FrameSVGNefrex.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameSVGNefrex = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importStar(require("react"));
const tools_1 = require("@arwes/tools");
const index_1 = require("../FrameSVG/index");
const toPath = (points) => points.map((p, i) => [i === 0 ? 'M' : 'L', ...p]);
const FrameSVGNefrex = (props) => {
    const { squareSize: ss = 16, strokeWidth = 1, smallLineLength: sll = 16, largeLineLength: lll = 64, padding: p = 0, className, ...otherProps } = props;
    const paths = (0, react_1.useMemo)(() => {
        const so = strokeWidth / 2; // Stroke offset.
        const polylineStyle = {
            stroke: 'currentcolor',
            strokeLinecap: 'round',
            strokeLinejoin: 'round',
            strokeWidth: String(strokeWidth),
            fill: 'none'
        };
        const leftTopLine = [
            [so + p, sll + ss + so + p],
            [so + p, ss + so + p],
            [ss + so + p, so + p],
            [ss + lll + so + p, so + p]
        ];
        const rightBottomLine = [
            [`100% - ${so + p}`, `100% - ${sll + ss + so + p}`],
            [`100% - ${so + p}`, `100% - ${ss + so + p}`],
            [`100% - ${ss + so + p}`, `100% - ${so + p}`],
            [`100% - ${ss + lll + so + p}`, `100% - ${so + p}`]
        ];
        const paths = [
            {
                name: 'bg',
                style: {
                    strokeWidth: 0,
                    fill: 'currentcolor'
                },
                path: toPath(leftTopLine
                    .concat([
                    [`100% - ${so + p}`, so + p]
                ])
                    .concat(rightBottomLine)
                    .concat([
                    [so + p, `100% - ${so + p}`]
                ]))
            },
            {
                name: 'line',
                style: polylineStyle,
                path: toPath(leftTopLine)
            },
            {
                name: 'line',
                style: polylineStyle,
                path: toPath(rightBottomLine)
            }
        ];
        return paths;
    }, [sll, lll, ss, strokeWidth, p]);
    return (react_1.default.createElement(index_1.FrameSVG, { ...otherProps, className: (0, tools_1.cx)('arwes-react-frames-framesvgnefrex', className), paths: paths }));
};
exports.FrameSVGNefrex = FrameSVGNefrex;
