// @arwes/react-frames/build/cjs/FrameSVGUnderline/FrameSVGUnderline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameSVGUnderline = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importStar(require("react"));
const tools_1 = require("@arwes/tools");
const index_1 = require("../FrameSVG/index");
const FrameSVGUnderline = (props) => {
    const { squareSize: ss = 16, strokeWidth: sw = 1, inverted, className, ...otherProps } = props;
    const paths = (0, react_1.useMemo)(() => {
        const so = sw / 2;
        return [
            {
                name: 'bg',
                style: {
                    strokeWidth: 0,
                    fill: 'currentcolor'
                },
                path: [
                    ['M', 0, 0],
                    ['L', 0, '100%'],
                    ['L', `100% - ${ss}`, '100%'],
                    ['L', '100%', `100% - ${ss}`],
                    ['L', '100%', 0]
                ]
            },
            {
                name: 'line',
                style: {
                    stroke: 'currentcolor',
                    strokeLinecap: 'round',
                    strokeLinejoin: 'round',
                    strokeWidth: String(sw),
                    fill: 'none'
                },
                path: [
                    ['M', so, `100% - ${so}`],
                    ['L', `100% - ${ss}`, `100% - ${so}`],
                    ['L', `100% - ${so}`, `100% - ${ss - so}`]
                ]
            }
        ];
    }, [ss, sw, inverted]);
    return (react_1.default.createElement(index_1.FrameSVG, { ...otherProps, className: (0, tools_1.cx)('arwes-react-frames-framesvgunderline', className), paths: paths }));
};
exports.FrameSVGUnderline = FrameSVGUnderline;
