// @arwes/react-animated/build/cjs/animations/animations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.aaVisibility = exports.aaOpacity = exports.aa = void 0;
const motion_1 = require("motion");
const aa = (prop, from, to, back) => ({
    transitions: {
        entering: { [prop]: [from, to] },
        exiting: { [prop]: [to, back !== null && back !== void 0 ? back : from] }
    }
});
exports.aa = aa;
const aaOpacity = () => aa('opacity', 0, 1);
exports.aaOpacity = aaOpacity;
const aaVisibility = () => {
    return {
        transitions: {
            entering: ({ element, duration }) => (0, motion_1.timeline)([
                [element, { opacity: [0, 1] }],
                [element, { opacity: [1, 0.7] }],
                [element, { opacity: [0.7, 1] }]
            ], { duration }),
            exiting: ({ element, duration }) => (0, motion_1.timeline)([
                [element, { opacity: [1, 0] }],
                [element, { opacity: [0, 0.3] }],
                [element, { opacity: [0.3, 0] }]
            ], { duration })
        }
    };
};
exports.aaVisibility = aaVisibility;
