// @arwes/react-frames/build/esm/FrameSVGLines/FrameSVGLines.js
import React, { useMemo } from 'react';
import { cx } from '@arwes/tools';
import { FrameSVG } from "../FrameSVG/index.js";
const FrameSVGLines = (props) => {
    const { largeLineWidth: llw = 1, smallLineWidth: slw = 1, smallLineLength: sll = 16, className, ...otherProps } = props;
    const paths = useMemo(() => {
        const polylineStyle = {
            strokeLinecap: 'square',
            stroke: 'currentcolor',
            fill: 'none'
        };
        const llo = llw / 2;
        const slo = slw / 2;
        const largePolylines = [
            // Top
            [
                ['M', llo, llo],
                ['L', '50% + 0.1', llo]
            ],
            [
                ['M', `100% - ${llo}`, llo],
                ['L', '50% - 0.1', llo]
            ],
            // Bottom
            [
                ['M', llo, `100% - ${llo}`],
                ['L', '50% + 0.1', `100% - ${llo}`]
            ],
            [
                ['M', `100% - ${llo}`, `100% - ${llo}`],
                ['L', '50% - 0.1', `100% - ${llo}`]
            ]
        ];
        const smallPolylines = [
            // Top
            [
                ['M', slo, llw + slo],
                ['L', sll + slo, llw + slo]
            ],
            [
                ['M', `100% - ${slo}`, llw + slo],
                ['L', `100% - ${sll + slo}`, llw + slo]
            ],
            // Bottom
            [
                ['M', slo, `100% - ${llw + slo}`],
                ['L', sll + slo, `100% - ${llw + slo}`]
            ],
            [
                ['M', `100% - ${slo}`, `100% - ${llw + slo}`],
                ['L', `100% - ${sll + slo}`, `100% - ${llw + slo}`]
            ]
        ];
        return [
            {
                name: 'bg',
                style: {
                    strokeWidth: 0,
                    fill: 'currentcolor'
                },
                path: [
                    ['M', 0, 0],
                    ['L', 0, '100%'],
                    ['L', '100%', '100%'],
                    ['L', '100%', 0]
                ]
            },
            ...largePolylines.map(polyline => ({
                name: 'line',
                style: {
                    ...polylineStyle,
                    strokeWidth: String(llw)
                },
                path: polyline
            })),
            ...smallPolylines.map(polyline => ({
                name: 'line',
                style: {
                    ...polylineStyle,
                    strokeWidth: String(slw)
                },
                path: polyline
            }))
        ];
    }, [llw, slw, sll]);
    return (React.createElement(FrameSVG, { ...otherProps, className: cx('arwes-react-frames-framesvglines', className), paths: paths }));
};
export { FrameSVGLines };
