// @arwes/react-frames/build/esm/FrameSVGUnderline/FrameSVGUnderline.js
import React, { useMemo } from 'react';
import { cx } from '@arwes/tools';
import { FrameSVG } from "../FrameSVG/index.js";
const FrameSVGUnderline = (props) => {
    const { squareSize: ss = 16, strokeWidth: sw = 1, inverted, className, ...otherProps } = props;
    const paths = useMemo(() => {
        const so = sw / 2;
        return [
            {
                name: 'bg',
                style: {
                    strokeWidth: 0,
                    fill: 'currentcolor'
                },
                path: [
                    ['M', 0, 0],
                    ['L', 0, '100%'],
                    ['L', `100% - ${ss}`, '100%'],
                    ['L', '100%', `100% - ${ss}`],
                    ['L', '100%', 0]
                ]
            },
            {
                name: 'line',
                style: {
                    stroke: 'currentcolor',
                    strokeLinecap: 'round',
                    strokeLinejoin: 'round',
                    strokeWidth: String(sw),
                    fill: 'none'
                },
                path: [
                    ['M', so, `100% - ${so}`],
                    ['L', `100% - ${ss}`, `100% - ${so}`],
                    ['L', `100% - ${so}`, `100% - ${ss - so}`]
                ]
            }
        ];
    }, [ss, sw, inverted]);
    return (React.createElement(FrameSVG, { ...otherProps, className: cx('arwes-react-frames-framesvgunderline', className), paths: paths }));
};
export { FrameSVGUnderline };
