// @arwes/react-frames/build/esm/FrameSVGCorners/FrameSVGCorners.js
import React, { useMemo } from 'react';
import { cx } from '@arwes/tools';
import { FrameSVG } from "../FrameSVG/index.js";
const FrameSVGCorners = (props) => {
    const { strokeWidth: cw = 1, cornerLength: cl = 16, className, ...otherProps } = props;
    const paths = useMemo(() => {
        const co = cw / 2;
        const bg = {
            name: 'bg',
            style: {
                strokeWidth: 0,
                fill: 'currentcolor'
            },
            path: [
                ['M', cw, cw],
                ['L', cw, `100% - ${cw}`],
                ['L', `100% - ${cw}`, `100% - ${cw}`],
                ['L', `100% - ${cw}`, cw]
            ]
        };
        const linesPaths = [
            // Left top.
            [['M', co, co], ['L', co, cl]],
            [['M', co, co], ['L', cl, co]],
            // Right top.
            [['M', `100% - ${co}`, co], ['L', `100% - ${cl}`, co]],
            [['M', `100% - ${co}`, co], ['L', `100% - ${co}`, cl]],
            // Right bottom.
            [['M', `100% - ${co}`, `100% - ${co}`], ['L', `100% - ${cl}`, `100% - ${co}`]],
            [['M', `100% - ${co}`, `100% - ${co}`], ['L', `100% - ${co}`, `100% - ${cl}`]],
            // Left bottom.
            [['M', co, `100% - ${co}`], ['L', co, `100% - ${cl}`]],
            [['M', co, `100% - ${co}`], ['L', cl, `100% - ${co}`]]
        ];
        const lines = linesPaths.map(path => ({
            name: 'line',
            style: {
                stroke: 'currentcolor',
                strokeLinecap: 'round',
                strokeLinejoin: 'round',
                strokeWidth: String(cw),
                fill: 'none'
            },
            path
        }));
        return [bg, ...lines];
    }, [cw, cl]);
    return (React.createElement(FrameSVG, { ...otherProps, className: cx('arwes-react-frames-framesvgcorners', className), paths: paths }));
};
export { FrameSVGCorners };
