// @arwes/react-frames/build/cjs/FrameSVGCorners/FrameSVGCorners.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameSVGCorners = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importStar(require("react"));
const tools_1 = require("@arwes/tools");
const index_1 = require("../FrameSVG/index");
const FrameSVGCorners = (props) => {
    const { strokeWidth: cw = 1, cornerLength: cl = 16, className, ...otherProps } = props;
    const paths = (0, react_1.useMemo)(() => {
        const co = cw / 2;
        const bg = {
            name: 'bg',
            style: {
                strokeWidth: 0,
                fill: 'currentcolor'
            },
            path: [
                ['M', cw, cw],
                ['L', cw, `100% - ${cw}`],
                ['L', `100% - ${cw}`, `100% - ${cw}`],
                ['L', `100% - ${cw}`, cw]
            ]
        };
        const linesPaths = [
            // Left top.
            [['M', co, co], ['L', co, cl]],
            [['M', co, co], ['L', cl, co]],
            // Right top.
            [['M', `100% - ${co}`, co], ['L', `100% - ${cl}`, co]],
            [['M', `100% - ${co}`, co], ['L', `100% - ${co}`, cl]],
            // Right bottom.
            [['M', `100% - ${co}`, `100% - ${co}`], ['L', `100% - ${cl}`, `100% - ${co}`]],
            [['M', `100% - ${co}`, `100% - ${co}`], ['L', `100% - ${co}`, `100% - ${cl}`]],
            // Left bottom.
            [['M', co, `100% - ${co}`], ['L', co, `100% - ${cl}`]],
            [['M', co, `100% - ${co}`], ['L', cl, `100% - ${co}`]]
        ];
        const lines = linesPaths.map(path => ({
            name: 'line',
            style: {
                stroke: 'currentcolor',
                strokeLinecap: 'round',
                strokeLinejoin: 'round',
                strokeWidth: String(cw),
                fill: 'none'
            },
            path
        }));
        return [bg, ...lines];
    }, [cw, cl]);
    return (react_1.default.createElement(index_1.FrameSVG, { ...otherProps, className: (0, tools_1.cx)('arwes-react-frames-framesvgcorners', className), paths: paths }));
};
exports.FrameSVGCorners = FrameSVGCorners;
