// @arwes/react-frames/build/cjs/FrameSVGLines/FrameSVGLines.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameSVGLines = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importStar(require("react"));
const tools_1 = require("@arwes/tools");
const index_1 = require("../FrameSVG/index");
const FrameSVGLines = (props) => {
    const { largeLineWidth: llw = 1, smallLineWidth: slw = 1, smallLineLength: sll = 16, className, ...otherProps } = props;
    const paths = (0, react_1.useMemo)(() => {
        const polylineStyle = {
            strokeLinecap: 'square',
            stroke: 'currentcolor',
            fill: 'none'
        };
        const llo = llw / 2;
        const slo = slw / 2;
        const largePolylines = [
            // Top
            [
                ['M', llo, llo],
                ['L', '50% + 0.1', llo]
            ],
            [
                ['M', `100% - ${llo}`, llo],
                ['L', '50% - 0.1', llo]
            ],
            // Bottom
            [
                ['M', llo, `100% - ${llo}`],
                ['L', '50% + 0.1', `100% - ${llo}`]
            ],
            [
                ['M', `100% - ${llo}`, `100% - ${llo}`],
                ['L', '50% - 0.1', `100% - ${llo}`]
            ]
        ];
        const smallPolylines = [
            // Top
            [
                ['M', slo, llw + slo],
                ['L', sll + slo, llw + slo]
            ],
            [
                ['M', `100% - ${slo}`, llw + slo],
                ['L', `100% - ${sll + slo}`, llw + slo]
            ],
            // Bottom
            [
                ['M', slo, `100% - ${llw + slo}`],
                ['L', sll + slo, `100% - ${llw + slo}`]
            ],
            [
                ['M', `100% - ${slo}`, `100% - ${llw + slo}`],
                ['L', `100% - ${sll + slo}`, `100% - ${llw + slo}`]
            ]
        ];
        return [
            {
                name: 'bg',
                style: {
                    strokeWidth: 0,
                    fill: 'currentcolor'
                },
                path: [
                    ['M', 0, 0],
                    ['L', 0, '100%'],
                    ['L', '100%', '100%'],
                    ['L', '100%', 0]
                ]
            },
            ...largePolylines.map(polyline => ({
                name: 'line',
                style: {
                    ...polylineStyle,
                    strokeWidth: String(llw)
                },
                path: polyline
            })),
            ...smallPolylines.map(polyline => ({
                name: 'line',
                style: {
                    ...polylineStyle,
                    strokeWidth: String(slw)
                },
                path: polyline
            }))
        ];
    }, [llw, slw, sll]);
    return (react_1.default.createElement(index_1.FrameSVG, { ...otherProps, className: (0, tools_1.cx)('arwes-react-frames-framesvglines', className), paths: paths }));
};
exports.FrameSVGLines = FrameSVGLines;
