// @arwes/react-frames/build/cjs/FrameSVGOctagon/FrameSVGOctagon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameSVGOctagon = void 0;
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importStar(require("react"));
const tools_1 = require("@arwes/tools");
const index_1 = require("../FrameSVG/index");
const toPath = (points) => points.map((p, i) => [i === 0 ? 'M' : 'L', ...p]);
const FrameSVGOctagon = (props) => {
    const { leftTop = true, rightTop = true, rightBottom = true, leftBottom = true, squareSize: ss = 16, strokeWidth = 1, padding: p = 0, className, ...otherProps } = props;
    const paths = (0, react_1.useMemo)(() => {
        const so = strokeWidth / 2;
        const polylineStyle = {
            stroke: 'currentcolor',
            strokeLinecap: 'round',
            strokeLinejoin: 'round',
            strokeWidth: String(strokeWidth),
            fill: 'none'
        };
        const leftTopPoints = leftTop
            ? [
                [ss + so + p, so + p],
                [so + p, ss + so + p]
            ]
            : [
                [so + p, so + p]
            ];
        const leftBottomPoints = leftBottom
            ? [
                [so + p, `100% - ${ss + p}`],
                [ss + so + p, `100% - ${so + p}`]
            ]
            : [
                [so + p, `100% - ${so + p}`]
            ];
        const rightBottomPoints = rightBottom
            ? [
                [`100% - ${ss + so + p}`, `100% - ${so + p}`],
                [`100% - ${so + p}`, `100% - ${ss + so + p}`]
            ]
            : [
                [`100% - ${so + p}`, `100% - ${so + p}`]
            ];
        const rightTopPoints = rightTop
            ? [
                [`100% - ${so + p}`, ss - so + p],
                [`100% - ${ss - so + p}`, so + p]
            ]
            : [
                [`100% - ${so + p}`, so + p]
            ];
        // leftTop > leftBottom > rightBottom
        const polyline1 = toPath([
            ...leftTopPoints,
            ...leftBottomPoints,
            rightBottomPoints[0]
        ]);
        // rightBottom > rightTop > leftTop
        const polyline2 = toPath([
            ...rightBottomPoints,
            ...rightTopPoints,
            leftTopPoints[0]
        ]);
        const paths = [
            {
                name: 'bg',
                style: {
                    strokeWidth: 0,
                    fill: 'currentcolor'
                },
                path: polyline1.concat(polyline2)
            },
            {
                name: 'line',
                style: polylineStyle,
                path: polyline1
            },
            {
                name: 'line',
                style: polylineStyle,
                path: polyline2
            }
        ];
        return paths;
    }, [leftTop, rightTop, rightBottom, leftBottom, ss, strokeWidth, p]);
    return (react_1.default.createElement(index_1.FrameSVG, { ...otherProps, className: (0, tools_1.cx)('arwes-react-frames-framesvgoctagon', className), paths: paths }));
};
exports.FrameSVGOctagon = FrameSVGOctagon;
